sympy/solvers/ode/ode.py

```diff
@@ -1014,45 +1014,28 @@ class in it.  Note that a hint may do this anyway if
-    # integrals e.g.:
-    # d^3/dx^3(x y) = F(x)
     ode = SingleODEProblem(eq_orig, func, x, prep=prep, xi=xi, eta=eta)
-    solvers = {
-        NthAlgebraic: ('nth_algebraic',),
-        FirstExact:('1st_exact',),
-        FirstLinear: ('1st_linear',),
-        AlmostLinear: ('almost_linear',),
-        Bernoulli: ('Bernoulli',),
-        Factorable: ('factorable',),
-        RiccatiSpecial: ('Riccati_special_minus2',),
-        SecondNonlinearAutonomousConserved: ('2nd_nonlinear_autonomous_conserved',),
-        Liouville: ('Liouville',),
-        Separable: ('separable',),
-        SeparableReduced: ('separable_reduced',),
-        HomogeneousCoeffSubsDepDivIndep: ('1st_homogeneous_coeff_subs_dep_div_indep',),
-        HomogeneousCoeffSubsIndepDivDep: ('1st_homogeneous_coeff_subs_indep_div_dep',),
-        HomogeneousCoeffBest: ('1st_homogeneous_coeff_best',),
-        LinearCoefficients: ('linear_coefficients',),
-        NthOrderReducible: ('nth_order_reducible',),
-        SecondHypergeometric: ('2nd_hypergeometric',),
-        NthLinearConstantCoeffHomogeneous: ('nth_linear_constant_coeff_homogeneous',),
-        NthLinearConstantCoeffVariationOfParameters: ('nth_linear_constant_coeff_variation_of_parameters',),
-        NthLinearConstantCoeffUndeterminedCoefficients: ('nth_linear_constant_coeff_undetermined_coefficients',),
-        NthLinearEulerEqHomogeneous: ('nth_linear_euler_eq_homogeneous',),
-        NthLinearEulerEqNonhomogeneousVariationOfParameters: ('nth_linear_euler_eq_nonhomogeneous_variation_of_parameters',),
-        NthLinearEulerEqNonhomogeneousUndeterminedCoefficients: ('nth_linear_euler_eq_nonhomogeneous_undetermined_coefficients',),
-        SecondLinearBessel: ('2nd_linear_bessel',),
-        SecondLinearAiry: ('2nd_linear_airy',),
-        LieGroup: ('lie_group',),
-    }
-    for solvercls in solvers:
-        solver = solvercls(ode)
+    user_hint = kwargs.get('hint','default')
+    # Used when dsolve is called without an explicit hint.
+    # We exit early to return the first valid match
+    early_exit = (user_hint=='default')
+    if user_hint.endswith('_Integral'):
+        user_hint = user_hint[:-len('_Integral')]
+    user_map = solver_map
+    # An explicit hint has been given to dsolve
+    # Skip matching code for other hints
+    if user_hint not in ['default', 'all', 'all_Integral', 'best'] and user_hint in solver_map:
+        user_map = {user_hint: solver_map[user_hint]}
+
+    for hint in user_map:
+        solver = user_map[hint](ode)
         if solver.matches():
-            for hints in solvers[solvercls]:
-                matching_hints[hints] = solver
-                if solvercls.has_integral:
-                    matching_hints[hints + "_Integral"] = solver
+            matching_hints[hint] = solver
+            if user_map[hint].has_integral:
+                matching_hints[hint + "_Integral"] = solver
+            if dict and early_exit:
+                break
```

Comment:
Why is this break rather than return?

Presumably it's because some of the other hints are not yet refactored and might come earlier than the match that we just found. Which hints are not refactored?

Looking at `allhints` it seems that everything is refactored apart from the series solvers. If that's the case then I think it might be better to just return here. I don't see why we would want to return a series solution when it is still possible to return a non-series solution. Note that the matching code for the series solvers can be slow.